spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DistriParameterSynchronizer.scala

```diff
@@ -0,0 +1,369 @@
+    ".fetchCompletionPoolSize", "1").toInt
+
+  private val asyncTaskWaitingQueue : PriorityBlockingQueue[AsyncFutureTask] =
+    new PriorityBlockingQueue[AsyncFutureTask](taskSize)
+
+  private val blockFetchRequestQueue: PriorityBlockingQueue[BlockFetchRequest] =
+    new PriorityBlockingQueue[BlockFetchRequest](taskSize)
+
+  // thread pool to remove expired blocks
+  private lazy val clearPool: ExecutorService =
+    Executors.newFixedThreadPool(clearPoolSize, new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  // main thread pool to do put-get-aggregate
+  private val workerPool: ExecutorService =
+    Executors.newFixedThreadPool(workerPoolSize, new ThreadFactory {
+    override def newThread(r: Runnable): Thread = {
+      val t = Executors.defaultThreadFactory().newThread(r)
+      t.setDaemon(true)
+      t
+    }
+  })
+
+  // long running thread to fetch the request
+  workerPool.submit(new Runnable {
+    override def run(): Unit = {
+      val asyncFutureTask = asyncTaskWaitingQueue.take
+      workerPool.submit(asyncFutureTask.task)
+    }
+  })
+
+  // thread pool for put and aggregate
+  private lazy val syncPool: ExecutorService = Executors.newFixedThreadPool(syncPoolSize,
+    new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  // thread pool for fetching blocks
+  private lazy val fetchPool: ExecutorService = Executors.newFixedThreadPool(syncPoolSize,
+    new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  // thread pool to update sow on fetching completion
+  private val fetchCompletionPool: ExecutorService = Executors.
+    newFixedThreadPool(fetchCompletionPoolSize,
+    new ThreadFactory {
+      override def newThread(r: Runnable): Thread = {
+        val t = Executors.defaultThreadFactory().newThread(r)
+        t.setDaemon(true)
+        t
+      }
+    })
+
+  (0 until syncPoolSize).foreach(th => {
+    fetchPool.submit(new Runnable {
+      override def run(): Unit = {
+        while (true && !shutdown) {
+          val fetchRequest = blockFetchRequestQueue.take
+          val syncMeta = fetchRequest.syncMeta
+          val pid = fetchRequest.futureTask.fetchOnCompletion.fromPartition
+          val parameterBlockId = getParameterBlockId(syncMeta.name,
+          syncMeta.counter, pid, partitionID)
+          val block = BlockManagerWrapper.getLocalOrRemoteBytes(parameterBlockId)
+          if (block == None) {
+            // promote the priporty in next fetch
+            fetchRequest.priority += 1
```

Comment:
why increase the priority if the data is not ready?

Reply:
every block is supposed to be fetched in the right order, so the one that was not fetched should be prioritized as it has highest possibility in next fetch